File: Ping/Backend/InvitationService.swift
import Foundation
import OSLog

private let signposter = OSSignposter(subsystem: "com.youngminpark.ping.Ping", category: "polling")

@MainActor
final class InvitationService {
    private let client: SupabaseClient

    init(client: SupabaseClient = .shared) {
        self.client = client
    }

    func send(fromUid: String, fromNickname: String, toUid: String, roomId: String, roomName: String) async throws {
        let displayName = RoomLimits.sanitizedRoomName(roomName)
        guard RoomLimits.isValidRoomName(displayName) else { throw PingError.invalidRoomName }

        let _: String = try await client.rpcValue("ping_send_invitation", body: [
            "to_uid": toUid,
            "room_uuid": roomId,
            "from_nickname": fromNickname,
            "room_name_text": displayName
        ])
    }

    @discardableResult
    func inviteUser(toUid: String, fromNickname: String, roomName: String) async throws -> Room {
        let displayName = RoomLimits.sanitizedRoomName(roomName)
        guard RoomLimits.isValidRoomName(displayName) else { throw PingError.invalidRoomName }

        let rooms: [Room] = try await client.rpcArray("ping_invite_user", body: [
            "target_uid": toUid,
            "inviter_nickname_text": fromNickname,
            "room_name_text": displayName,
            "searchable_room_name": SearchableText.normalize(displayName)
        ])

        guard let room = rooms.first else { throw PingError.supabaseUnavailable }
        return room
    }

    func observeIncoming(uid: String) -> AsyncStream<[Invitation]> {
        let client = self.client
        return AsyncStream { continuation in
            let task = Task.detached(priority: .utility) {
                while !Task.isCancelled {
                    let intervalState = signposter.beginInterval("invitations-poll-cycle")
                    do {
                        let invitations: [Invitation] = try await client.rpcArray("ping_incoming_invitations")
                        continuation.yield(invitations)
                    } catch {
                        continuation.yield([])
                    }
                    signposter.endInterval("invitations-poll-cycle", intervalState)
                    try? await Task.sleep(nanoseconds: 2_000_000_000)
                }

                continuation.finish()
            }

            continuation.onTermination = { _ in task.cancel() }
        }
    }

    func accept(invitation: Invitation, myUid: String, myNickname: String, roomService: RoomService) async throws {
        guard let inviteId = invitation.id else { return }

        try await client.rpcVoid("ping_accept_invitation", body: [
            "invitation_uuid": inviteId,
            "nickname_text": myNickname
        ])
    }

    func reject(inviteId: String) async throws {
        try await client.rpcVoid("ping_reject_invitation", body: [
            "invitation_uuid": inviteId
        ])
    }

    func createInviteLink(roomId: String) async throws -> InviteLink {
        let links: [InviteLink] = try await client.rpcArray("ping_create_invite_link", body: [
            "room_uuid": roomId
        ])

        guard let link = links.first else { throw PingError.supabaseUnavailable }
        return link
    }

    @discardableResult
    func acceptInviteLink(token: String, nickname: String) async throws -> Room {
        let rooms: [Room] = try await client.rpcArray("ping_accept_invite_link", body: [
            "invite_token": token,
            "nickname_text": nickname
        ])

        guard let room = rooms.first else { throw PingError.roomUnavailable }
        return room
    }
}

File: windows/src/Ping.Windows.Core/Backend/InvitationService.cs
using System.Text.Json.Serialization;
using Ping.Windows.Core.Models;

namespace Ping.Windows.Core.Backend;

public sealed class InvitationService(ISupabaseRpcClient client)
{
    public Task<string> SendAsync(
        string toUid,
        string roomId,
        string fromNickname,
        string roomName,
        CancellationToken cancellationToken = default) =>
        client.RpcValueAsync<string>(
            "ping_send_invitation",
            new SendInvitationRpcBody(toUid, roomId, fromNickname, RoomName.Normalize(roomName)),
            cancellationToken);

    public async Task<Room> InviteUserAsync(
        string targetUid,
        string inviterNickname,
        string roomName,
        CancellationToken cancellationToken = default)
    {
        var normalized = RoomName.Normalize(roomName);
        var rooms = await client.RpcArrayAsync<Room>(
            "ping_invite_user",
            new InviteUserRpcBody(
                targetUid,
                inviterNickname,
                normalized,
                SearchableText.Normalize(normalized)),
            cancellationToken).ConfigureAwait(false);
        return rooms.FirstOrDefault()
            ?? throw new InvalidOperationException("Invite-user RPC returned no room.");
    }

    public Task<IReadOnlyList<Invitation>> IncomingAsync(CancellationToken cancellationToken = default) =>
        client.RpcArrayAsync<Invitation>("ping_incoming_invitations", cancellationToken: cancellationToken);

    public Task AcceptAsync(
        string invitationId,
        string nickname,
        CancellationToken cancellationToken = default) =>
        client.RpcVoidAsync(
            "ping_accept_invitation",
            new AcceptInvitationRpcBody(invitationId, nickname),
            cancellationToken);

    public Task RejectAsync(string invitationId, CancellationToken cancellationToken = default) =>
        client.RpcVoidAsync(
            "ping_reject_invitation",
            new InvitationIdRpcBody(invitationId),
            cancellationToken);

    public async Task<InviteLink> CreateInviteLinkAsync(
        string roomId,
        CancellationToken cancellationToken = default)
    {
        var links = await client.RpcArrayAsync<InviteLink>(
            "ping_create_invite_link",
            new RoomIdRpcBody(roomId),
            cancellationToken).ConfigureAwait(false);
        return links.FirstOrDefault()
            ?? throw new InvalidOperationException("Invite-link RPC returned no link.");
    }

    public async Task<Room> AcceptInviteLinkAsync(
        string token,
        string nickname,
        CancellationToken cancellationToken = default)
    {
        var rooms = await client.RpcArrayAsync<Room>(
            "ping_accept_invite_link",
            new AcceptInviteLinkRpcBody(token, nickname),
            cancellationToken).ConfigureAwait(false);
        return rooms.FirstOrDefault()
            ?? throw new InvalidOperationException("Accept invite-link RPC returned no room.");
    }
}

public sealed record SendInvitationRpcBody(
    [property: JsonPropertyName("to_uid")] string ToUid,
    [property: JsonPropertyName("room_uuid")] string RoomUuid,
    [property: JsonPropertyName("from_nickname")] string FromNickname,
    [property: JsonPropertyName("room_name_text")] string RoomNameText);

public sealed record InviteUserRpcBody(
    [property: JsonPropertyName("target_uid")] string TargetUid,
    [property: JsonPropertyName("inviter_nickname_text")] string InviterNicknameText,
    [property: JsonPropertyName("room_name_text")] string RoomNameText,
    [property: JsonPropertyName("searchable_room_name")] string SearchableRoomName);

public sealed record AcceptInvitationRpcBody(
    [property: JsonPropertyName("invitation_uuid")] string InvitationUuid,
    [property: JsonPropertyName("nickname_text")] string NicknameText);

public sealed record InvitationIdRpcBody(
    [property: JsonPropertyName("invitation_uuid")] string InvitationUuid);

public sealed record AcceptInviteLinkRpcBody(
    [property: JsonPropertyName("invite_token")] string InviteToken,
    [property: JsonPropertyName("nickname_text")] string NicknameText);
